File: paicord-rs/PaicordLibBridge/Sources/DiscordHTTP/Types/Channel.swift
import PaicordLib

class BridgedDiscordChannel {
    var channel: DiscordChannel

    init(channel: DiscordChannel) {
        self.channel = channel
    }

    func get_id() -> UInt64 {
        return UInt64(self.channel.id.rawValue) ?? 0
    }

    func has_guild_id() -> Bool {
        return self.channel.guild_id != nil
    }

    func get_guild_id() -> SnowflakeRust {
        return SnowflakeRust(inner: UInt64(self.channel.guild_id!.rawValue) ?? 0)
    }

    func get_type() -> DiscordChannelKindRust {
        if let kind = self.channel.type {
            switch kind {
            case .guildText:
                return .GuildText
            case .dm:
                return .Dm
            case .guildVoice:
                return .GuildVoice
            case .groupDm:
                return .GroupDm
            case .guildCategory:
                return .GuildCategory
            case .guildAnnouncement:
                return .GuildAnnouncement
            case .guildStore:
                return .GuildStore
            case .announcementThread:
                return .AnnouncementThread
            case .publicThread:
                return .PublicThread
            case .privateThread:
                return .PrivateThread
            case .guildStageVoice:
                return .GuildStageVoice
            case .guildDirectory:
                return .GuildDirectory
            case .guildForum:
                return .GuildForum
            case .guildMedia:
                return .GuildMedia
            case .__undocumented(_):
                return .Undocumented
            }
        } else {
            return .Undocumented
        }
    }

    func get_name() -> Optional<String> {
        return self.channel.name
    }

    func get_topic() -> Optional<String> {
        return self.channel.topic
    }

    func get_parent_id() -> Optional<UInt64> {
        if let parent_id = self.channel.parent_id {
            return UInt64(parent_id.rawValue)
        } else {
            return nil
        }
    }

    func get_position() -> Optional<Int32> {
        if let position = self.channel.position {
            return Int32(position)
        } else {
            return nil
        }
    }
}

class BridgedPartialMessage {
    var message: DiscordChannel.PartialMessage

    init(message: DiscordChannel.PartialMessage) {
        self.message = message
    }

    func get_kind() -> DiscordMessageKindRust {
        if let kind: DiscordChannel.Message.Kind = self.message.type {
            switch kind {
                case .`default`:  // 0
                    return .Default
                case .recipientAdd:  // 1
                    return .RecipientAdd
                case .recipientRemove:  // 2
                    return .RecipientRemove
                case .call:  // 3
                    return .Call
                case .channelNameChange:  // 4
                    return .ChannelNameChange
                case .channelIconChange:  // 5
                    return .ChannelIconChange
                case .channelPinnedMessage:  // 6
                    return .ChannelPinnedMessage
                case .guildMemberJoin:  // 7
                    return .GuildMemberJoin
                case .userPremiumGuildSubscription:  // 8
                    return .UserPremiumGuildSubscription
                case .userPremiumGuildSubscriptionTier1:  // 9
                    return .UserPremiumGuildSubscriptionTier1
                case .userPremiumGuildSubscriptionTier2:  // 10
                    return .UserPremiumGuildSubscriptionTier2
                case .userPremiumGuildSubscriptionTier3:  // 11
                    return .UserPremiumGuildSubscriptionTier3
                case .channelFollowAdd:  // 12
                    return .ChannelFollowAdd
                case .guildDiscoveryDisqualified:  // 14
                    return .GuildDiscoveryDisqualified
                case .guildDiscoveryRequalified:  // 15
                    return .GuildDiscoveryRequalified
                case .guildDiscoveryGracePeriodInitialWarning:  // 16
                    return .GuildDiscoveryGracePeriodInitialWarning
                case .guildDiscoveryGracePeriodFinalWarning:  // 17
                    return .GuildDiscoveryGracePeriodFinalWarning
                case .threadCreated:  // 18
                    return .ThreadCreated
                case .reply:  // 19
                    return .Reply
                case .chatInputCommand:  // 20
                    return .ChatInputCommand
                case .threadStarterMessage:  // 21
                    return .ThreadStarterMessage
                case .guildInviteReminder:  // 22
                    return .GuildInviteReminder
                case .contextMenuCommand:  // 23
                    return .ContextMenuCommand
                case .autoModerationAction:  // 24
                    return .AutoModerationAction
                case .roleSubscriptionPurchase:  // 25
                    return .RoleSubscriptionPurchase
                case .interactionPremiumUpsell:  // 26
                    return .InteractionPremiumUpsell
                case .stageStart:  // 27
                    return .StageStart
                case .stageEnd:  // 28
                    return .StageEnd
                case .stageSpeaker:  // 29
                    return .StageSpeaker
                case .stageRaiseHand:  // 30
                    return .StageRaiseHand
                case .stageTopic:  // 31
                    return .StageTopic
                case .guildApplicationPremiumSubscription:  // 32
                    return .GuildApplicationPremiumSubscription
                case .premiumReferral:  // 35
                    return .PremiumReferral
                case .guildIncidentAlertModeEnabled:  // 36
                    return .GuildIncidentAlertModeEnabled
                case .guildIncidentAlertModeDisabled:  // 37
                    return .GuildIncidentAlertModeDisabled
                case .guildIncidentReportRaid:  // 38
                    return .GuildIncidentReportRaid
                case .guildIncidentReportFalseAlarm:  // 39
                    return .GuildIncidentReportFalseAlarm
                case .guildDeadchatRevivePrompt:  // 40
                    return .GuildDeadchatRevivePrompt
                case .customGift:  // 41
                    return .CustomGift
                case .guildGamingStatsPrompt:  // 42
                    return .GuildGamingStatsPrompt
                case .purchaseNotification:  // 44
                    return .PurchaseNotification
                case .pollResult:  // 46
                    return .PollResult
                case .changelog:  // 47
                    return .Changelog
                case .nitroNotification:  // 48
                    return .NitroNotification
                case .channelLinkedToLobby:  // 49
                    return .ChannelLinkedToLobby
                case .giftingPrompt:  // 50
                    return .GiftingPrompt
                case .inGameMessageNux:  // 51
                    return .InGameMessageNux
                case .guildJoinRequestAcceptNotification:  // 52
                    return .GuildJoinRequestAcceptNotification
                case .guildJoinRequestRejectNotification:  // 53
                    return .GuildJoinRequestRejectNotification
                case .guildJoinRequestWithdrawnNotification:  // 54
                    return .GuildJoinRequestWithdrawnNotification
                case .hdStreamingUpgraded:  // 55
                    return .HdStreamingUpgraded
                case .reportToModDeletedMessage:  // 58
                    return .ReportToModDeletedMessage
                case .reportToModTimeoutUser:  // 59
                    return .ReportToModTimeoutUser
                case .reportToModKickUser:  // 60
                    return .ReportToModKickUser
                case .reportToModBanUser:  // 61
                    return .ReportToModBanUser
                case .reportToModClosedReport:  // 62
                    return .ReportToModClosedReport
                case .emojiAdded:  // 63
                    return .EmojiAdded
                case .__undocumented(_):
                    return .Default
            }
        } else {
            return DiscordMessageKindRust.Default
        }
    }

    func get_id() -> SnowflakeRust {
        return SnowflakeRust(inner: UInt64(self.message.id.rawValue) ?? 0)
    }

    func get_channel_id() -> SnowflakeRust {
        return SnowflakeRust(inner: UInt64(self.message.channel_id.rawValue) ?? 0)
    }

    func has_guild_id() -> Bool {
        return self.message.guild_id != nil
    }

    func get_guild_id() -> SnowflakeRust {
        return SnowflakeRust(inner: UInt64(self.message.guild_id!.rawValue) ?? 0)
    }

    func has_author() -> Bool {
        return self.message.author != nil
    }

    func get_author() -> DiscordUserRust {
        return DiscordUserRust(user: self.message.author!)
    }

    func get_content() -> Optional<String> {
        return self.message.content
    }

    func has_timestamp() -> Bool {
        return self.message.timestamp != nil
    }

    func get_timestamp() -> DiscordTimestampRust {
        return DiscordTimestampRust(timestamp: self.message.timestamp!)
    }

    func has_edited_timestamp() -> Bool {
        return self.message.edited_timestamp != nil
    }

    func get_edited_timestamp() -> DiscordTimestampRust {
        return DiscordTimestampRust(timestamp: self.message.edited_timestamp!)
    }

    func attachment_count() -> Optional<UInt> {
        if let attachments = self.message.attachments {
            return UInt(attachments.count)
        } else {
            return nil
        }
    }

    func get_attachment(index: UInt) -> AttachmentRust {
        return AttachmentRust(attachment: self.message.attachments![Int(index)])
    }

    func has_referenced_message() -> Bool {
        return self.message.referenced_message != nil
    }

    func get_referenced_message() -> BridgedPartialMessage {
        BridgedPartialMessage(message: self.message.referenced_message!.value)
    }

    func has_member() -> Bool {
        return self.message.member != nil
    }

    func get_member() -> BridgedPartialMember {
        BridgedPartialMember(member: self.message.member!)
    }
}

extension AttachmentRust {
    init(attachment: DiscordChannel.Message.Attachment) {
        self.id = SnowflakeRust(inner: UInt64(attachment.id.rawValue) ?? 0)
        self.filename = attachment.filename.intoRustString()
        self.title = attachment.title?.intoRustString() ?? nil
        self.description = attachment.description?.intoRustString() ?? nil
        self.content_type = attachment.content_type?.intoRustString() ?? nil
        self.size = Int32(attachment.size)
        self.url = attachment.url.intoRustString()
        self.proxy_url = attachment.proxy_url.intoRustString()
        self.placeholder = attachment.placeholder?.intoRustString() ?? nil
        self.height = attachment.height.map({ Int32($0) })
        self.width = attachment.width.map({ Int32($0) })
        self.ephemeral = attachment.ephemeral
        self.duration_secs = attachment.duration_secs
        self.waveform = attachment.waveform?.intoRustString() ?? nil
    }
}

File: paicord-rs/src/discord_models/types/channel.rs
use crate::{discord_models::types::snowflake::Snowflake, ffi};

pub type DiscordChannelKind = crate::ffi::DiscordChannelKindRust;
pub type Attachment = crate::ffi::AttachmentRust;
pub type DiscordMessageKind = crate::ffi::DiscordMessageKindRust;

#[derive(Clone, Debug)]
pub struct DiscordChannel {
    pub id: Snowflake,
    pub guild_id: Option<Snowflake>,
    pub name: Option<String>,
    pub topic: Option<String>,
    pub kind: Option<DiscordChannelKind>,
    pub parent_id: Option<Snowflake>,
    pub position: Option<i32>,
}

impl DiscordChannel {
    pub fn new(inner: crate::ffi::BridgedDiscordChannel) -> Self {
        Self {
            id: Snowflake::from(inner.get_id()),
            guild_id: if inner.has_guild_id() {
                Some(Snowflake::from(inner.get_guild_id()))
            } else {
                None
            },
            name: inner.get_name(),
            topic: inner.get_topic(),
            kind: if inner.get_type() == DiscordChannelKind::Undocumented {
                None
            } else {
                Some(inner.get_type())
            },
            parent_id: inner.get_parent_id().map(Snowflake::from),
            position: inner.get_position(),
        }
    }
}

impl std::fmt::Debug for DiscordChannelKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::GuildText => write!(f, "GuildText"),
            Self::Dm => write!(f, "Dm"),
            Self::GuildVoice => write!(f, "GuildVoice"),
            Self::GroupDm => write!(f, "GroupDm"),
            Self::GuildCategory => write!(f, "GuildCategory"),
            Self::GuildAnnouncement => write!(f, "GuildAnnouncement"),
            Self::GuildStore => write!(f, "GuildStore"),
            Self::AnnouncementThread => write!(f, "AnnouncementThread"),
            Self::PublicThread => write!(f, "PublicThread"),
            Self::PrivateThread => write!(f, "PrivateThread"),
            Self::GuildStageVoice => write!(f, "GuildStageVoice"),
            Self::GuildDirectory => write!(f, "GuildDirectory"),
            Self::GuildForum => write!(f, "GuildForum"),
            Self::GuildMedia => write!(f, "GuildMedia"),
            Self::Undocumented => write!(f, "Undocumented"),
        }
    }
}

impl PartialEq for DiscordChannelKind {
    fn eq(&self, other: &Self) -> bool {
        matches!(
            (self, other),
            (Self::GuildText, Self::GuildText)
                | (Self::Dm, Self::Dm)
                | (Self::GuildVoice, Self::GuildVoice)
                | (Self::GroupDm, Self::GroupDm)
                | (Self::GuildCategory, Self::GuildCategory)
                | (Self::GuildAnnouncement, Self::GuildAnnouncement)
                | (Self::GuildStore, Self::GuildStore)
                | (Self::AnnouncementThread, Self::AnnouncementThread)
                | (Self::PublicThread, Self::PublicThread)
                | (Self::PrivateThread, Self::PrivateThread)
                | (Self::GuildStageVoice, Self::GuildStageVoice)
                | (Self::GuildDirectory, Self::GuildDirectory)
                | (Self::GuildForum, Self::GuildForum)
                | (Self::GuildMedia, Self::GuildMedia)
                | (Self::Undocumented, Self::Undocumented)
        )
    }
}

impl Clone for Attachment {
    fn clone(&self) -> Self {
        Self {
            id: self.id.clone(),
            filename: self.filename.clone(),
            title: self.title.clone(),
            description: self.description.clone(),
            content_type: self.content_type.clone(),
            size: self.size.clone(),
            url: self.url.clone(),
            proxy_url: self.proxy_url.clone(),
            placeholder: self.placeholder.clone(),
            height: self.height.clone(),
            width: self.width.clone(),
            ephemeral: self.ephemeral.clone(),
            duration_secs: self.duration_secs.clone(),
            waveform: self.waveform.clone(),
        }
    }
}

impl std::fmt::Debug for Attachment {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Attachment")
            .field("id", &self.id)
            .field("filename", &self.filename)
            .field("title", &self.title)
            .field("description", &self.description)
            .field("content_type", &self.content_type)
            .field("size", &self.size)
            .field("url", &self.url)
            .field("proxy_url", &self.proxy_url)
            .field("placeholder", &self.placeholder)
            .field("height", &self.height)
            .field("width", &self.width)
            .field("ephemeral", &self.ephemeral)
            .field("duration_secs", &self.duration_secs)
            .field("waveform", &self.waveform)
            .finish()
    }
}

impl std::fmt::Debug for DiscordMessageKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Default => write!(f, "Default"),
            Self::RecipientAdd => write!(f, "RecipientAdd"),
            Self::RecipientRemove => write!(f, "RecipientRemove"),
            Self::Call => write!(f, "Call"),
            Self::ChannelNameChange => write!(f, "ChannelNameChange"),
            Self::ChannelIconChange => write!(f, "ChannelIconChange"),
            Self::ChannelPinnedMessage => write!(f, "ChannelPinnedMessage"),
            Self::GuildMemberJoin => write!(f, "GuildMemberJoin"),
            Self::UserPremiumGuildSubscription => write!(f, "UserPremiumGuildSubscription"),
            Self::UserPremiumGuildSubscriptionTier1 => write!(f, "UserPremiumGuildSubscriptionTier1"),
            Self::UserPremiumGuildSubscriptionTier2 => write!(f, "UserPremiumGuildSubscriptionTier2"),
            Self::UserPremiumGuildSubscriptionTier3 => write!(f, "UserPremiumGuildSubscriptionTier3"),
            Self::ChannelFollowAdd => write!(f, "ChannelFollowAdd"),
            Self::GuildDiscoveryDisqualified => write!(f, "GuildDiscoveryDisqualified"),
            Self::GuildDiscoveryRequalified => write!(f, "GuildDiscoveryRequalified"),
            Self::GuildDiscoveryGracePeriodInitialWarning => write!(f, "GuildDiscoveryGracePeriodInitialWarning"),
            Self::GuildDiscoveryGracePeriodFinalWarning => write!(f, "GuildDiscoveryGracePeriodFinalWarning"),
            Self::ThreadCreated => write!(f, "ThreadCreated"),
            Self::Reply => write!(f, "Reply"),
            Self::ChatInputCommand => write!(f, "ChatInputCommand"),
            Self::ThreadStarterMessage => write!(f, "ThreadStarterMessage"),
            Self::GuildInviteReminder => write!(f, "GuildInviteReminder"),
            Self::ContextMenuCommand => write!(f, "ContextMenuCommand"),
            Self::AutoModerationAction => write!(f, "AutoModerationAction"),
            Self::RoleSubscriptionPurchase => write!(f, "RoleSubscriptionPurchase"),
            Self::InteractionPremiumUpsell => write!(f, "InteractionPremiumUpsell"),
            Self::StageStart => write!(f, "StageStart"),
            Self::StageEnd => write!(f, "StageEnd"),
            Self::StageSpeaker => write!(f, "StageSpeaker"),
            Self::StageRaiseHand => write!(f, "StageRaiseHand"),
            Self::StageTopic => write!(f, "StageTopic"),
            Self::GuildApplicationPremiumSubscription => write!(f, "GuildApplicationPremiumSubscription"),
            Self::PremiumReferral => write!(f, "PremiumReferral"),
            Self::GuildIncidentAlertModeEnabled => write!(f, "GuildIncidentAlertModeEnabled"),
            Self::GuildIncidentAlertModeDisabled => write!(f, "GuildIncidentAlertModeDisabled"),
            Self::GuildIncidentReportRaid => write!(f, "GuildIncidentReportRaid"),
            Self::GuildIncidentReportFalseAlarm => write!(f, "GuildIncidentReportFalseAlarm"),
            Self::GuildDeadchatRevivePrompt => write!(f, "GuildDeadchatRevivePrompt"),
            Self::CustomGift => write!(f, "CustomGift"),
            Self::GuildGamingStatsPrompt => write!(f, "GuildGamingStatsPrompt"),
            Self::PurchaseNotification => write!(f, "PurchaseNotification"),
            Self::PollResult => write!(f, "PollResult"),
            Self::Changelog => write!(f, "Changelog"),
            Self::NitroNotification => write!(f, "NitroNotification"),
            Self::ChannelLinkedToLobby => write!(f, "ChannelLinkedToLobby"),
            Self::GiftingPrompt => write!(f, "GiftingPrompt"),
            Self::InGameMessageNux => write!(f, "InGameMessageNux"),
            Self::GuildJoinRequestAcceptNotification => write!(f, "GuildJoinRequestAcceptNotification"),
            Self::GuildJoinRequestRejectNotification => write!(f, "GuildJoinRequestRejectNotification"),
            Self::GuildJoinRequestWithdrawnNotification => write!(f, "GuildJoinRequestWithdrawnNotification"),
            Self::HdStreamingUpgraded => write!(f, "HdStreamingUpgraded"),
            Self::ReportToModDeletedMessage => write!(f, "ReportToModDeletedMessage"),
            Self::ReportToModTimeoutUser => write!(f, "ReportToModTimeoutUser"),
            Self::ReportToModKickUser => write!(f, "ReportToModKickUser"),
            Self::ReportToModBanUser => write!(f, "ReportToModBanUser"),
            Self::ReportToModClosedReport => write!(f, "ReportToModClosedReport"),
            Self::EmojiAdded => write!(f, "EmojiAdded"),
        }
    }
}

unsafe impl Send for ffi::BridgedMessageVec {}
unsafe impl Sync for ffi::BridgedMessageVec {}